packages/x-data-grid/src/hooks/features/pagination/useGridRowCount.ts

```diff
@@ -123,8 +161,35 @@ export const useGridRowCount = (
   React.useEffect(() => {
     if (props.paginationMode === 'client') {
       apiRef.current.setRowCount(visibleTopLevelRowCount);
-    } else if (props.rowCount != null) {
+      return;
+    }
+
+    if (isNumber(props.rowCount) && prevRowCountProp.current !== props.rowCount) {
+      prevRowCountProp.current = props.rowCount;
       apiRef.current.setRowCount(props.rowCount);
+      return;
+    }
+
+    const isLastPage = paginationMeta.hasNextPage === false;
+    if (
+      isLastPage &&
+      (rowCountState === -1 || rowCountState % paginationModel.pageSize === 0) &&
+      visibleTopLevelRowCount < paginationModel.pageSize
+    ) {
+      // Actual count of the rows is less than the page size, reflect the value
+      apiRef.current.setRowCount(
+        (rowCountState !== -1
+          ? rowCountState - paginationModel.pageSize
+          : paginationModel.pageSize * paginationModel.page) + visibleTopLevelRowCount,
+      );
```

Comment:
Shouldn't this be simpler?
```suggestion
    if (isLastPage && rowCountState === -1) {
      apiRef.current.setRowCount(paginationModel.pageSize * paginationModel.page) + visibleTopLevelRowCount);
```

And maybe keeping `handlePaginationMetaChange` would make more sense?
See https://github.com/mui/mui-x/pull/12474#discussion_r1553347579

Reply:
After reading https://github.com/mui/mui-x/pull/12474#discussion_r1553368477, I understand that `handlePaginationMetaChange` won't work.
Perhaps add a separate `useEffect` to handle unknown row count specifically?

```ts
  React.useEffect(() => {
    if (props.paginationMode === 'client') {
      apiRef.current.setRowCount(visibleTopLevelRowCount);
    } else if (props.rowCount != null) {
      apiRef.current.setRowCount(props.rowCount);
    }
  }, [apiRef, visibleTopLevelRowCount, props.paginationMode, props.rowCount]);
  
  const isLastPage = paginationMeta.hasNextPage === false;
  React.useEffect(() => {
    if (isLastPage && rowCountState === -1) {
      apiRef.current.setRowCount(paginationModel.pageSize * paginationModel.page) + visibleTopLevelRowCount);
    }
  }, [apiRef, isLastPage, rowCountState, paginationModel, visibleTopLevelRowCount])
```

Reply:
Yeah, let's do it. I didn't do it previously because many dependencies were common, but this does seem better. Thanks!